[docs] Development quickguide

tasosalvas opened the pull request:
## Developers getting started
>Not affiliated with any sort of video productions.

So, this PR does 3 things, in 3 separate commits so you can cherrypick.

### Added a section about running `make test`
Do tell me if my writing style is too much. I'm happy with it, but it's your project's voice and I'll be happy to adapt.

### I went wild on the thing
I broke the guide into sections, just so it's more scannable when used as a reference.

I also renamed the thing and changed its starting line. I mean, it doesn't really describe the development process -- and isn't everything a "draft" in the grand scheme of things? :)

The code section on syncing with upstream was repeated twice, so I replaced it with a sentence stating that it is the same process.

### Added `DEVELOPMENT.rst` and `CONTRIBUTING.rst` on the index
I think they both belong right on the top of the Developer Guide menu.
* `CONTRIBUTING.rst` was only referenced in the old documentation under "scripts".
* `DEVELOPMENT.rst` was completely loose.

Meanwhile the only content the Developer Guide section has is a descriptive text about CI and the DEPs, which are pretty intimidating.

I think both texts are also worth keeping in the project root instead of moving into the documentation, so even the wildest of cowboys can find them.

---

I've been trying to get my head around the documentation's structure, to see how else I can help. I see the docs seem to be in the middle of a migration. And those empty placeholder pages make the main menu feel like a minefield.

A document describing what is to be done would be pretty handy (maybe "contibuting to the documentation" or something). The placeholder page titles could go live there as well.

But I digress. Tell me what you think and if you want any changes.

tasosalvas committed:
[docs] Add `make test` section to DEVELOPMENT.rst

tasosalvas committed:
[docs] Aggresively restructure DEVELOPMENT.rst

tasosalvas committed:
[docs] Add CONTRIBUTING.rst and DEVELOPMENT.rst to the doc index

drybjed commented on DEVELOPMENT.rst:
```diff
@@ -111,14 +120,50 @@ reflect new commits.
+   sudo apt install build-essential git graphviz shellcheck
+   pip install -U pycodestyle nose2 sphinx yamllint ansible-lint
+
+You also need **Docker**, and a user able to run docker containers.
```
Currently Docker is not really required per-se - the only thing that needs it is verification of the `Dockerfile` and checking if Ansible and DebOps run fin inside of a Docker container. The test code in `make test` checks if Docker is available, and if not, this particular test is successfully skipped.

Although I now wonder if the entire `make test` should be done inside Docker as well, would that make life of the developers easier? Although at least documentation would probably have to be built locally to be readable, otherwise the container would have to have a webserver. Hmm, food for thought.

tasosalvas replied:
Oh, I had a bad installation of docker when I started, so I guess I confused docker's error message with an unmet dependency.

Personally I'd like it if docker wasn't involved at all in the first round of tests, since it looks like it hits the debian repos for a full upgrade every time it runs. Sounds more fitting for a proper testing environment with an apt mirror.

Also uh, having to uninstall it to skip it doesn't really make it optional.

I think `make docs` and `make docker` might be worth mentioning in the documentation. Would it make sense to change `make lint` to include all the linters and document that as well?

At any rate, I think I should mention `make help` in this guide.

drybjed replied:
I suppose that the Docker test could be removed from `test-all` in the Makefile, and then called specifically on Travis-CI. I'll update that when I get back.

drybjed reviewed (commented):
Very nice work, thanks! One thing I would change is to make the Docker dependency in test suite optional, see the comment below.

I think that you are right with all of those empty sections. I hoped that the documentation would be written sooner, but as usual other things took precedence - I guess I just like writing roles more than documentation. Your idea about folding the empty sections for the moment and putting the desired documentation structure in some kind of development documentation is a good one, want to give it a spin in this PR? I'm currently on holidays, will be back in the next week so you have some time to write it.

tasosalvas commented:
No worries, answer at your convenience and enjoy your holidays.

I will have a go at cleaning up the empty sections and creating a contributing to the doc guide page, yeah.

I should get back to building my own infra too - I was thinking I might grill the irc for advice and maybe flesh out a few of the empty "admins" sections. I always keep notes while I work so it's not too much extra effort to produce some doc, if I know what's needed. 

Oh, should I remove `CONTRIBUTING.rst` from the old doc now that it's in the new?

tasosalvas committed:
[docs] Moved placeholder content into `contributing-docs.rst`

New file `developer-guide/contributing-docs.rst` contains:
- A call to action
- A brief overview of how the docs work
- A list of doc subject ideas, which used to clutter the index

tasosalvas commented:
## Folded placeholder content into `contributing-docs.rst`
Wow I hadn't realized `development-model.rst` was drawn by graphviz at build-time. Cool. 

### Created `developer-guide/contributing-docs.rst`
I've written a quick intro/guide and listed all of the removed pages as contribution ideas.

"Project roadmap" I think doesn't really belong there. Looks more like a note-to-self. I left it for completeness, but tell me if I should remove it.


### Removed all placeholder pages
The **Tester guide** section is completely gone! My opinion is that the deleted subjects would be better as sections of a `developer-guide/testing.rst`, at least until they have enough content to warrant their own section.

I don't think "testers" is as robust as the rest of the use-case distinctions, anyway.

---

At any rate, I should adapt `overview.rst` before this PR is merged, since it now refers to a non-existent thing.

tasosalvas commented:
## Should I migrate the Old Documentation?
Ok now I get it, these are before monorepo times.

Lol I've studied this doc way more than any sane person should. :)

I think I can rehome everything in the current documentation. Perhaps in a more correctly named PR? I could do it here as well if the branch name is no big deal.

There's good content in here that I totally missed in my first reading of the doc.

### debops-tools
* `introduction.rst`: Broad DebOps overview. I think it's redundant, delete
* `configuration.rst`: Should be `User guide` -> `Script configuration`
* `installation.rst`: Outdated version of `/INSTALL.rst`, delete
* `scripts/*`: Should be `User guide` -> `The DebOps CLI`

### debops-playbooks
* `introduction.rst`: meh, delete
* `playbook-layout.rst`: `User guide`
* `custom-features.rst`: Could be merged with `User guide` -> `DebOps for Ansible users`
* `custom-environment.rst`: `User guide`
* `guides/*`: `Admin guide`

### debops-policy
This probably deserves its own section without the "Old" designation.

It would also help the "Introduction" section breathe a little if we moved a few stuff like "Philosophy", "Timeline" and "References" to such a meta-section.

* `introduction.rst`: doesn't say much, delete
* `organizational-structure.rst`
* `distributed-development.rst`
* `code-signing-policy.rst`
* `security-policy.rst`
* `planned-policy-changes.rst`
* `references.rst`: a worse version of `introduction/references.rst`, delete
* `changelog.rst`: changelog for the `debops-policy` repo. delete?
* `COPYRIGHT`/`copyright.rst`: Is the monorepo's enough, or does this have to be kept? It's in the commit history anyway.

---

Some of those might need some checking from more experienced DebOps users than me to ensure they're not out of date. Anyone feel free to chime in! But anyway, I think doing this will be a step in the right direction and if something wrong ends up in the doc it can be fixed in place later.

drybjed commented:
> I think I can rehome everything in the current documentation. Perhaps in a more correctly named PR? I could do it here as well if the branch name is no big deal.

You can do it here if you want, branch names aren't really that important. I'll try to backport everything to the stable releases as well, as long as there are no merge conflicts.

> There's good content in here that I totally missed in my first reading of the doc.
> ### debops-tools
> 
>     * `introduction.rst`: Broad DebOps overview. I think it's redundant, delete
> 
>     * `configuration.rst`: Should be `User guide` -> `Script configuration`
> 
>     * `installation.rst`: Outdated version of `/INSTALL.rst`, delete
> 
>     * `scripts/*`: Should be `User guide` -> `The DebOps CLI`
> 
> 
> ### debops-playbooks
> 
>     * `introduction.rst`: meh, delete
> 
>     * `playbook-layout.rst`: `User guide`
> 
>     * `custom-features.rst`: Could be merged with `User guide` -> `DebOps for Ansible users`
> 
>     * `custom-environment.rst`: `User guide`
> 
>     * `guides/*`: `Admin guide`
> 
> 
> ### debops-policy
> 
> This probably deserves its own section without the "Old" designation.
> 
> It would also help the "Introduction" section breathe a little if we moved a few stuff like "Philosophy", "Timeline" and "References" to such a meta-section.
> 
>     * `introduction.rst`: doesn't say much, delete
> 
>     * `organizational-structure.rst`
> 
>     * `distributed-development.rst`
> 
>     * `code-signing-policy.rst`
> 
>     * `security-policy.rst`
> 
>     * `planned-policy-changes.rst`
> 
>     * `references.rst`: a worse version of `introduction/references.rst`, delete
> 
>     * `changelog.rst`: changelog for the `debops-policy` repo. delete?
> 
>     * `COPYRIGHT`/`copyright.rst`: Is the monorepo's enough, or does this have to be kept? It's in the commit history anyway.

All good suggestions, keep it up. Sure, the copyrigt can be removed, it was relevant when the policy was in its own repository.

> Some of those might need some checking from more experienced DebOps users than me to ensure they're not out of date. Anyone feel free to chime in! But anyway, I think doing this will be a step in the right direction and if something wrong ends up in the doc it can be fixed in place later.

Agreed. Thanks again for taking care of the documentation! :-)

tasosalvas committed:
[docs] Kill the old docs with fire

Cleaned debops-tools:
- `configuration.rst` and `scripts/*` now live in `user-guide`
- A chunk of `introduction.rst` ended up in `scripts.rst`
- Removed two items from the doc suggestions on `contributing-docs.rst`

Cleaned debops-playbooks:
- `playbook-layout.rst` is now `user-guide/playbooks.rst`.
- `custom-environment.rst` is now `user-guide/custom-environment.rst`
- `guides/*` are now 3 new pages in `admin-guide`
- `custom-features.rst` merged with `user-guide/debops-for-ansible.rst`

Cleaned debops-policy:
- It now lives in meta/policy
- A bunch of cruft files from the old repo deleted

Reorganized the index:
- Removed the old docs section and the relevant note.
- Created a new `meta` category, for things like policy and philosophy

tasosalvas commented:
## The *kill it with fire* upgrade
Uh, is there some way to see the sidebar menu in the rendered docs? Am I missing something? All my index placements are guesses and the structure has changed a ton from the live version, so I'm kinda working blind at this point.

I tried to keep editing to the minimum needed to repurpose the documents, like changing titles to fit the new structure and removing references to the old repos. That said, I couldn't help myself with making minor grammar and formatting edits all over the place.

Oh and in the end I reordered the menu a bit.

A few random points:
* Should `user-guide/playbooks.rst` mention the bootstrap playbook as well?
* On `debops-for-ansible.rst` I think the merge went ok, but do check.
* Should I remove "planned-policy-changes", since its only planned change refers to a past date?
* Does "The debops api" maybe belong inside the new "The DebOps project" menu?

I would like `user-guide/scripts.rst` to hold a one-line description for each of the commands, and then link to the specific command page; So that the reader can get an overview without digging into the files, but the links on the sidebar still point to the individual documents. Gotta read a bit more sphinx/rst to be sure how to do that, though.

What's the deal with `.. include:: ../includes/global.rst`? I see it in some documents but not in all. I'll be happy to enforce one way or the other once I know which way it should be. A ton of the refs it defines are for the old repos, btw.

Still gotta fix that mention of docker and update `overview.rst`, but the bulk of the work is done.

drybjed commented:
> Uh, is there some way to see the sidebar menu in the rendered docs? Am I missing something? All my index placements are guesses and the structure has changed a ton from the live version, so I'm kinda working blind at this point.

Yes, on Debian you can install `python3-sphinx-rtd-theme` along with `python3-sphinx` to get the ReadTheDocs.org theme available. It will be used automatically if present.

> I tried to keep editing to the minimum needed to repurpose the documents, like changing titles to fit the new structure and removing references to the old repos. That said, I couldn't help myself with making minor grammar and formatting edits all over the place.

That's greaf, fix away. :-) English is not my native language and my keyboard sometimes gets stuck and I miss the error. Unfortunately, documentation cannot be executed to check the syntax and correctness - maybe that should be a thing already? Imagine unit and integration tests for English...

>     * Should `user-guide/playbooks.rst` mention the bootstrap playbook as well?

I think that there should be a separate section for playbooks, similar to current section for roles. I don't mean there to describe every single role playbook, but the main ones (`site.yml`, `common.yml`, `bootstrap*.yml`, the `ldap/init-directory.yml`, any new ones that show up). Each significant playbook could have its own page with reasons behind it, included roles, etc.

>     * On `debops-for-ansible.rst` I think the merge went ok, but do check.

Will do. I'll render the PR soon to see it in action.

>     * Should I remove "planned-policy-changes", since its only planned change refers to a past date?

Yes, I think it is time. Code signing is currently kind of enforced by GitHub, but I'm merging all of the pull reuests manually anyway so the signed status doesn't matter for GitHub. I think that the project is too small to enforce code signing from everyone that wants to contribute, at least at this point. For now each change is properly reviewed by me before merging. Merge commits are also signed to authenticate them against my GPG key.

>     * Does "The debops api" maybe belong inside the new "The DebOps project" menu?

Sure, you can move it there.

> I would like `user-guide/scripts.rst` to hold a one-line description for each of the commands, and then link to the specific command page; So that the reader can get an overview without digging into the files, but the links on the sidebar still point to the individual documents. Gotta read a bit more sphinx/rst to be sure how to do that, though.

Sounds good.

> What's the deal with `.. include:: ../includes/global.rst`? I see it in some documents but not in all. I'll be happy to enforce one way or the other once I know which way it should be. A ton of the refs it defines are for the old repos, btw.

reStructuredText enforces all named links to be universal across the entire set of documents. Because in the old days each role had its own documentation in its own repository, which then were merged in a separate repository via `git submodule` entries. To ease the management burden of synchronizing all the links across the repositories, I think @ypid came up with `includes/global.rst` which would have them all in one place and included at the build time.

Now, all the documentation is in one place and it should be easier to spot the named link duplicates during documentation tests, but the old documentation hasn't been updated so the `includes/global.rst` is still there. These days I usually use anonymous links which can have duplicates, so I feel that having a central place for common set of links is kind of awkward, especially that there's a separate script to update the links (for example, add new Ansible module documentation) and generate new `includes/global.rst`. But since all of the references are still in place, it will have to stay for now.

Speaking of which, run `make links`. :-)

tasosalvas commented:
> Yes, on Debian you can install `python3-sphinx-rtd-theme` along with `python3-sphinx` to get the ReadTheDocs.org theme available. It will be used automatically if present.

Ah! Thank you! That is much better; and useful for the `contributing-docs` guide.

> I think that there should be a separate section for playbooks, similar to current section for roles.

I see the need for a hierarchy with playbooks, as it will make easier to know what to update when changing the playbooks themselves, but currently all the doc we have is essentially derived from the comments in `site.yml`. I think a single good doc example would be better than laying placeholders again.

The primary thing I would find useful would be a document that helps users find the partial playbook they want to run. This could probably be accomplished in a section of `playbooks`, though.

"The Site playbook" is confusing, btw, it should be "Overriding the site playbook", sitting alongside "Playbooks", or being part of "Project directory".

I really like the use-case approach of the users/admins/developers categories. So I was, on the contrary, thinking about shoving roles in user-guide. :) I mean, roles are not distinct from DebOps anymore, and I spend like 95% of my user time there. (I'm only talking about menu entries, of course, not the dir.)

Introduction is pretty empty now, so `installation` could go live there, and the User guide's title could become `User manual`, to signify that the content there is meant to be descriptive of all user-facing features (as opposed to procedural guides, like installation).

Or there could be a `[User] reference` section, encompassing both playbooks and roles, and that can be moved right next to the user guide on the top level menu. It wouldn't make the hierarchy any flatter, though.

I was also thinking `Admin recipes` would be a cute name and make that section's goals clearer at a glance.

> Now, all the documentation is in one place and it should be easier to spot the named link duplicates during documentation tests, but the old documentation hasn't been updated so the `includes/global.rst` is still there. These days I usually use anonymous links which can have duplicates, so I feel that having a central place for common set of links is kind of awkward, especially that there's a separate script to update the links (for example, add new Ansible module documentation) and generate new `includes/global.rst`. But since all of the references are still in place, it will have to stay for now.

Yeah it crossed my mind when I saw it that it would make casual contributions to the docs a bit more awkward.

So, the action to completely solve this would be to remove the includes and then fix existing refs until the doc builds again?

If I add clear directions in `contributing-docs`, then contributors can know to fix this over time as they make edits to random docs -- it shouldn't be very hard to write a script to do it, either.

> Speaking of which, run `make links`. :-)

Oh dang that's a lot of output. I should fix the ones in the files I handled.

It also could go in `contributing-docs` as a fun activity for everyone, something like:

run `make links`, look at `docs/_build/linkcheck/output.txt` then:
* Broken links should be fixed
* Permanent redirects should be pointed to the new addresses
* Temporary redirects are fine and the listed links are usually desirable

---

Okay, I have a bunch of obvious stuff to do according to what we already discussed.

I'll do a best effort attempt at getting `scripts/*` to look good (now that I can see what I'm doing). I'm not if sure my idea about the sidebar was good, anyway. I'll probably do a manual list like in the role index.

I'll also have a go at making `user-guide/playbooks.rst` show its contents properly in the sidebar too. I don't think we should include a playbooks overhaul as part of this PR, as it essentially involves writing a bunch of new content.

drybjed commented:
> I see the need for a hierarchy with playbooks, as it will make easier to know what to update when changing the playbooks themselves, but currently all the doc we have is essentially derived from the comments in `site.yml`. I think a single good doc example would be better than laying placeholders again.
> 
> The primary thing I would find useful would be a document that helps users find the partial playbook they want to run. This could probably be accomplished in a section of `playbooks`, though.

Sure, although apart from the main `srv`, `app`, etc. playbooks there are only bootstrap, common, role playbooks and some LDAP-related ones. But that probably could be expanded in the future.

> "The Site playbook" is confusing, btw, it should be "Overriding the site playbook", sitting alongside "Playbooks", or being part of "Project directory".

The "Project directory" would fit best with the context of overriding the site playbook, since you probably would want to do it there.

> I really like the use-case approach of the users/admins/developers categories. So I was, on the contrary, thinking about shoving roles in user-guide. :) I mean, roles are not distinct from DebOps anymore, and I spend like 95% of my user time there. (I'm only talking about menu entries, of course, not the dir.)

Agreed.

> Introduction is pretty empty now, so `installation` could go live there, and the User guide's title could become `User manual`, to signify that the content there is meant to be descriptive of all user-facing features (as opposed to procedural guides, like installation).
> 
> Or there could be a `[User] reference` section, encompassing both playbooks and roles, and that can be moved right next to the user guide on the top level menu. It wouldn't make the hierarchy any flatter, though.

Sure, "User manual" sounds good.

> I was also thinking `Admin recipes` would be a cute name and make that section's goals clearer at a glance.

Also a good idea.

> So, the action to completely solve this would be to remove the includes and then fix existing refs until the doc builds again?
> 
> If I add clear directions in `contributing-docs`, then contributors can know to fix this over time as they make edits to random docs -- it shouldn't be very hard to write a script to do it, either.

Yes, but that's not really a problem for now, so it can be handled over time.

> I'll also have a go at making `user-guide/playbooks.rst` show its contents properly in the sidebar too. I don't think we should include a playbooks overhaul as part of this PR, as it essentially involves writing a bunch of new content.

I think that this PR should be merged sooner rather than later, so that any other potential changes don't have merge conflicts. And we could get feedback from users about the changes.

tasosalvas committed:
[docs] Main menu organization, structural edits

Changed the names of quite a few menu items to make the structure more apparent
- "User guide" is now "User manual"
- "Admin guide" is now "Admin recipes"
- The Roles pages are now "Roles (by category)" and "Roles (index)"
- Removed the word "DebOps" from top level titles where it was redundant

Consolidated "User manual" as the one place for user reference
- `install.rst` now lives in `introduction`
- Roles are now listed in "User manual"

Structural edits
- Removed `planned-policy-changes.rst`
- `user-guide/playbooks.rst` and `scripts.rst` now show their contents on the sidebar
- `users/site-playbook.rst` is now a section in `project-directories.rst`
- Fixed a few broken links and references in the migrated content

New stuff
- Rewrote `overview.rst` to reflect the menu changes
- Created `testing.rst`, documenting `make` usage

tasosalvas committed:
[docs] Moved `Debops OID Registry` from `admin-guide` to `meta`

tasosalvas commented:
I *think* these last commits address all the changes we've talked about. I hope I didn't miss something, but hey, it builds, and I bet it fixes more typos than it introduces. :)

I made a `testing.rst` and moved `make` usage there, I hope you like. I fixed the docker phrasing there. I documented how `docker` runs with `test` if installed, if that changes, that's the place to mention it. 

I've also added some of the things we've talked about in `contributing-docs` as best practices.

I moved `Debops OID Registry` from `Admin recipes` to `meta`. I know nothing about ldap (yet!), I did it on a separate commit so you can ignore it and merge the others if I'm wrong.

There are a bunch of dead links in `meta/policy`, pointing to a `debops-keyring`. I had no idea what to do with them, so I left them as they are.

drybjed commented:
The `debops-keyring` was a separate `git` repository that held the GPG keys of DebOps developers and contributors. It sort-of made sense when DebOps was in separate `git` repositories to ensure that you can validate authenticity of a given repository. Now that everything is in one place, and it's much easier to fork the project, having a separate special place for GPG keys isn't neede anymore. These links could probably be removed.

I'm currently building the PR to check it out. Do you think it's ready for merge?

tasosalvas commented:
I think these are all the changes I had in mind, and I did my best to double-check everything is as intended.

If you don't see anything you would like done differently, yeah it's good to merge.

drybjed commented:
And, merged. @tasosalvas, really huge thanks for overhauling the documentation, I'm sure it wasn't an easy task! :-)
